src/platform/cmx/chal/cpuid.h

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2012 by Qi Wang, interwq@gwu.edu; Gabriel Parmer,
+ * gparmer@gwu.edu
+
+ * Redistribution of this file is permitted under the GNU General
+ * Public License v2.
+ */
+
+#ifndef CPUID_H
+#define CPUID_H
+
+#include "../../../kernel/include/chal/consts.h"
+#include "../../../kernel/include/user/cos_types.h"
+#include "../../../kernel/include/asm_ipc_defs.h"
+#include "../../../kernel/include/list.h"
+#include "../chal_asm_inc.h"
+
+static inline cycles_t
+tsc(void)
+{
+	unsigned long long ret;
+	rdtscll(ret);
+	//__asm__ __volatile__("rdtsc" : "=A" (ret));
+
+	return ret;
+}
+
+struct cos_cpu_local_info {
+	/*
+	 * orig_sysenter_esp SHOULD be the first variable here. The
+	 * sysenter interposition path gets tss from it.
+	 */
+	unsigned long *orig_sysenter_esp; /* 4Points to the end of the tss struct in Linux. This is not used when we port to CMX, just here for compatibility */
+	/***********************************************/
+	/* info saved in kernel stack for fast access. 4+4+4+12+8+8+8+4+4+4*/
+	unsigned long cpuid;
+	void       *curr_thd;
```

Comment:
A lot of this data is used as a "cache" for data to avoid pipeline stalls during the invocation path.  Much of that is not very useful when the limiting factor is just # of instructions executed.  Not sure if we should (eventually, not now) optimize here.